### PR TITLE
Prevent starting background workers with NOLOGIN

### PR DESCRIPTION
**Pull request by WireBaron**
This change will check sql commands which start a background worker
on a hypertable to verify that the table owner has permission to
log into the database.  This is necessary, as background workers for
these commands will run with the permissions of the table owner, and
thus immediately fail if unable to log in.

### CODE REVIEW

File: tsl/src/bgw_policy/drop_chunks_api.c [hunk 1]
```diff
@@ -72,6 +75,21 @@ drop_chunks_add_policy(PG_FUNCTION_ARGS)
...
+	if (!rform->rolcanlogin)
+	{
+		ReleaseSysCache(role_tup);
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_AUTHORIZATION_SPECIFICATION),
+				 errmsg("Unable to start drop chunks background process for hypertabletable \"%s\"",
+						hypertable->fd.table_name.data),
+				 errhint("Hypertable owner must have LOGIN permission to run background tasks")));
+	}
+	ReleaseSysCache(role_tup);
```
Comment (mkindahl) [inline tsl/src/bgw_policy/drop_chunks_api.c]: This is repeated three times with slight variations. Probably a good candidate for a separate function to do the job.

File: tsl/src/bgw_policy/compress_chunks_api.c [hunk 2]
```diff
@@ -92,6 +97,21 @@ compress_chunks_add_policy(PG_FUNCTION_ARGS)
...
+	role_tup = SearchSysCache1(AUTHOID, ObjectIdGetDatum(owner_id));
+	rform = (Form_pg_authid) GETSTRUCT(role_tup);
+
+	if (!rform->rolcanlogin)
+	{
+		ts_cache_release(hcache);
+		ReleaseSysCache(role_tup);
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_AUTHORIZATION_SPECIFICATION),
+				 errmsg("Unable to start compression background process for hypertabletable \"%s\"",
```
Comment (erimatnor) [inline tsl/src/bgw_policy/compress_chunks_api.c]:
```suggestion
				 errmsg("unable to start compression background process for hypertable \"%s\"",
```

I am wondering if this message shouldn't just say:

`"permission denied to create compression job for hypertable \"%s\""`

This gives a bit more context on why it failed and also seems more consistent with permission-related PG errors. (They usually are a variation of `permission denied to...`)


Note that these nits and suggestions apply to other places where similar errors occur as well.

File: tsl/src/bgw_policy/compress_chunks_api.c [hunk 3]
```diff
@@ -92,6 +97,21 @@ compress_chunks_add_policy(PG_FUNCTION_ARGS)
...
+						hypertable->fd.table_name.data),
+				 errhint("Hypertable owner must have LOGIN permission to run background tasks")));
```
Comment (erimatnor) [inline tsl/src/bgw_policy/compress_chunks_api.c]:
```suggestion
				 errhint("Hypertable owner must have LOGIN permission to run background tasks.")));
```

Comment (erimatnor) [reply tsl/src/bgw_policy/compress_chunks_api.c]: Same here.

File: tsl/src/bgw_policy/compress_chunks_api.c (hunk 2 above)
Comment (WireBaron) [reply tsl/src/bgw_policy/compress_chunks_api.c]: Went with `"permission denied to start ... background process"` after refactoring the code to a single location.

File: tsl/src/bgw_policy/drop_chunks_api.c (hunk 1 above)
Comment (WireBaron) [reply tsl/src/bgw_policy/drop_chunks_api.c]: Done.

File: tsl/src/bgw_policy/compress_chunks_api.c (hunk 3 above)
Comment (WireBaron) [reply tsl/src/bgw_policy/compress_chunks_api.c]: Fixed.

File: src/bgw/job.c [hunk 4]
```diff
@@ -451,6 +453,24 @@ ts_bgw_job_permission_check(BgwJob *job)
...
+	HeapTuple role_tup = SearchSysCache1(AUTHOID, ObjectIdGetDatum(owner));
+	Form_pg_authid rform = (Form_pg_authid) GETSTRUCT(role_tup);
+
+	if (!rform->rolcanlogin)
+	{
+		ReleaseSysCache(role_tup);
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_AUTHORIZATION_SPECIFICATION),
+				 errmsg("permission denied to start %s background process", job_type_names[type]),
+				 errhint("Hypertable owner must have LOGIN permission to run background tasks.")));
```
Comment (mkindahl) [inline src/bgw/job.c]: Minor: Supporting this is easier if the role name is shown in the error message.

File: src/bgw/job.c [hunk 5]
```diff
@@ -451,6 +453,24 @@ ts_bgw_job_permission_check(BgwJob *job)
 				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
 				 errmsg("insufficient permssions to alter job %d", job->fd.id)));
```
Comment (erimatnor) [inline src/bgw/job.c]:
```suggestion
				 errmsg("insufficient permissions to alter job %d", job->fd.id)));
```

Noticed a typo here. Realized not in your code, but seems worthwhile to fix (here or in other PR).

File: tsl/test/expected/compression_bgw.out [hunk 6]
```diff
@@ -185,3 +186,26 @@ order by chunk_name;
  test_table_int  | _timescaledb_internal._hyper_3_6_chunk | 24 kB                    | 16 kB
 (2 rows)
 
+--TEST 8
+--hypertable owner lacks permission to start background worker
+\c :TEST_DBNAME :ROLE_SUPERUSER
```
Comment (gayyappan) [inline tsl/test/expected/compression_bgw.out]: tests usually connect wit DEFAULT_PERM_ROLE_USER unless super user privileges are required. I don't think it is needed in this case. Could you please use one of the regular users instead?

**Pull request merged**